File: src/main/java/com/github/unchama/itemstackbuilder/AbstractItemStackBuilder.kt
package com.github.unchama.itemstackbuilder

import com.github.unchama.itemstackbuilder.component.IconComponent
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta

/**
 * ItemStackBuilderのベースとなる抽象クラス.
 *
 * @param T 派生クラス自身の型
 * @param M 派生クラスが生成する[ItemStack]が持つべきであろう[ItemMeta]の型.
 *
 * @author karayuu
 */
@Suppress("UNCHECKED_CAST")
abstract class AbstractItemStackBuilder<T : AbstractItemStackBuilder<T, M>, M: ItemMeta>
protected constructor(material: Material, durability: Short) : ItemStackBuilder {

  private val component: IconComponent = IconComponent(material, durability)

  override fun title(title: String): T {
    this.component.title = title
    return this as T
  }

  override fun lore(lore: List<String>): T {
    this.component.lore = lore
    return this as T
  }

  override fun enchanted(): T {
    this.component.isEnchanted = true
    return this as T
  }

  override fun unbreakable(): T {
    this.component.isUnbreakable = true
    return this as T
  }

  override fun amount(amount: Int): T {
    this.component.amount = amount
    return this as T
  }

  final override fun build(): ItemStack = component.itemStack.apply {
    itemMeta = (component.itemMeta as M).also { transformItemMetaOnBuild(it) }
  }

  /**
   * 生成されるアイテムスタックに入る[ItemMeta]を, ビルダー内の情報に基づいて変更する.
   */
  protected abstract fun transformItemMetaOnBuild(meta: M)
}

File: src/main/scala/com/github/unchama/itemstackbuilder/AbstractItemStackBuilder.scala
package com.github.unchama.itemstackbuilder

import com.github.unchama.itemstackbuilder.component.IconComponent
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.inventory.{ItemFlag, ItemStack}

import scala.collection.mutable

/**
 * ItemStackBuilderのベースとなる抽象クラス.
 *
 * @tparam M 派生クラスが生成する[ItemStack]が持つべきであろう[ItemMeta]の型.
 * @author karayuu
 */
abstract class AbstractItemStackBuilder[-M <: ItemMeta] protected
(material: Material, durability: Short) extends ItemStackBuilder {

  private val component: IconComponent = new IconComponent(material, durability)
  protected val enchants: mutable.Map[Enchantment, Int] = new mutable.HashMap()

  final override def build(): ItemStack = {
    val itemStack = component.itemStack()

    itemStack.setItemMeta {
      val meta = component.itemMeta().asInstanceOf[M]
      transformItemMetaOnBuild(meta)
      meta
    }

    itemStack
  }

  override def title(title: String): this.type = {
    this.component.title = title
    this
  }

  override def lore(lore: List[String]): this.type = {
    this.component.lore = lore
    this
  }

  override def enchanted(): this.type = {
    this.component.isEnchanted = true
    this
  }

  override def unbreakable(): this.type = {
    this.component.isUnbreakable = true
    this
  }

  override def amount(amount: Int): this.type = {
    this.component.amount = amount
    this
  }

  override def flagged(flag: ItemFlag): this.type = {
    this.component.itemFlagSet = component.itemFlagSet + flag
    this
  }

  /**
   * 生成されるアイテムスタックに入る[ItemMeta]を, ビルダー内の情報に基づいて変更する.
   */
  protected def transformItemMetaOnBuild(meta: M): Unit = {
    enchants.foreach { case (ench, lv) => meta.addEnchant(ench, lv, true) }
  }

  final def addEnchant(enchant: Enchantment, level: Int): this.type = {
    enchants(enchant) = level
    this
  }
}
